File: src/node/stress/stress_client.js
/*
 *
 * Copyright 2016, Google Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above
 * copyright notice, this list of conditions and the following disclaimer
 * in the documentation and/or other materials provided with the
 * distribution.
 *     * Neither the name of Google Inc. nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

'use strict';

var _ = require('lodash');

var grpc = require('../../..');

var interop_client = require('../interop/interop_client');
var MetricsServer = require('./metrics_server');

var running;

var metrics_server;

var start_time;
var query_count;

function makeCall(client, test_cases) {
  if (!running) {
    return;
  }
  var test_case = test_cases[_.random(test_cases.length - 1)];
  interop_client.test_cases[test_case].run(client, function() {
    query_count += 1;
    makeCall(client, test_cases);
  });
}

function makeCalls(client, test_cases, parallel_calls_per_channel) {
  _.times(parallel_calls_per_channel, function() {
    makeCall(client, test_cases);
  });
}

function getQps() {
  var diff = process.hrtime(start_time);
  var seconds = diff[0] + diff[1] / 1e9;
  return {long_value: query_count / seconds};
}

function start(server_addresses, test_cases, channels_per_server,
               parallel_calls_per_channel, metrics_port) {
  running = true;
  /* Assuming that we are not calling unimplemented_method. The client class
   * used by empty_unary is (currently) the client class used by every interop
   * test except unimplemented_method */
  var Client = interop_client.test_cases.empty_unary.Client;
  /* Make channels_per_server clients connecting to each server address */
  var channels = _.flatten(_.times(
      channels_per_server, _.partial(_.map, server_addresses, function(address) {
        return new Client(address, grpc.credentials.createInsecure());
      })));
  metrics_server = new MetricsServer(metrics_port);
  metrics_server.registerGauge('qps', getQps);
  start_time = process.hrtime();
  query_count = 0;
  _.each(channels, _.partial(makeCalls, _, test_cases,
                             parallel_calls_per_channel));
  metrics_server.start();
}

function stop() {
  running = false;
  metrics_server.shutdown();
  console.log('QPS: ' + getQps().long_value);
}

function main() {
  var parseArgs = require('minimist');
  var argv = parseArgs(process.argv, {
    string: ['server_addresses', 'test_cases', 'metrics_port'],
    default: {'server_addresses': 'localhost:8080',
              'test_duration_secs': -1,
              'num_channels_per_server': 1,
              'num_stubs_per_channel': 1,
              'metrics_port': '8081'}
  });
  var server_addresses = argv.server_addresses.split(',');
  /* Generate an array of test cases, where the number of instances of each name
   * corresponds to the number given in the argument.
   * e.g. 'empty_unary:1,large_unary:2' =>
   * ['empty_unary', 'large_unary', 'large_unary'] */
  var test_cases = _.flatten(_.map(argv.test_cases.split(','), function(value) {
    var split = value.split(':');
    return _.times(split[1], _.constant(split[0]));
  }));
  start(server_addresses, test_cases, argv.num_channels_per_server,
        argv.num_stubs_per_channel, argv.metrics_port);
  if (argv.test_duration_secs > -1) {
    setTimeout(stop, argv.test_duration_secs * 1000);
  }
}

main();

File: src/ruby/stress/stress_client.rb
#!/usr/bin/env ruby

# Copyright 2016 gRPC authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'optparse'
require 'thread'
require_relative '../pb/test/client'
require_relative './metrics_server'
require_relative '../lib/grpc'

class QpsGauge < Gauge
  @query_count
  @query_mutex
  @start_time

  def initialize
    @query_count = 0
    @query_mutex = Mutex.new
    @start_time = Time.now
  end

  def increment_queries
    @query_mutex.synchronize { @query_count += 1}
  end

  def get_name
    'qps'
  end

  def get_type
    'long'
  end

  def get_value
    (@query_mutex.synchronize { @query_count / (Time.now - @start_time) }).to_i
  end
end

def start_metrics_server(port)
  host = "0.0.0.0:#{port}"
  server = GRPC::RpcServer.new
  server.add_http2_port(host, :this_port_is_insecure)
  service = MetricsServiceImpl.new
  server.handle(service)
  server_thread = Thread.new { server.run_till_terminated }
  [server, service, server_thread]
end

StressArgs = Struct.new(:server_addresses, :test_cases, :duration,
                        :channels_per_server, :concurrent_calls, :metrics_port)

def start(stress_args)
  running = true
  threads = []
  qps_gauge = QpsGauge.new
  metrics_server, metrics_service, metrics_thread =
    start_metrics_server(stress_args.metrics_port)
  metrics_service.register_gauge(qps_gauge)
  stress_args.server_addresses.each do |address|
    stress_args.channels_per_server.times do
      client_args = Args.new
      client_args.host, client_args.port = address.split(':')
      client_args.secure = false
      client_args.test_case = ''
      stub = create_stub(client_args)
      named_tests = NamedTests.new(stub, client_args)
      stress_args.concurrent_calls.times do
        threads << Thread.new do
          while running
            named_tests.method(stress_args.test_cases.sample).call
            qps_gauge.increment_queries
          end
        end
      end
    end
  end
  if stress_args.duration >= 0
    sleep stress_args.duration
    running = false
    metrics_server.stop
    p "QPS: #{qps_gauge.get_value}"
    threads.each { |thd| thd.join; }
  end
  metrics_thread.join
end

def parse_stress_args
  stress_args = StressArgs.new
  stress_args.server_addresses = ['localhost:8080']
  stress_args.test_cases = []
  stress_args.duration = -1
  stress_args.channels_per_server = 1
  stress_args.concurrent_calls = 1
  stress_args.metrics_port = '8081'
  OptionParser.new do |opts|
    opts.on('--server_addresses [LIST]', Array) do |addrs|
      stress_args.server_addresses = addrs
    end
    opts.on('--test_cases cases', Array) do |cases|
      stress_args.test_cases = (cases.map do |item|
                                  split = item.split(':')
                                  [split[0]] * split[1].to_i
                                end).reduce([], :+)
    end
    opts.on('--test_duration_secs [INT]', OptionParser::DecimalInteger) do |time|
      stress_args.duration = time
    end
    opts.on('--num_channels_per_server [INT]', OptionParser::DecimalInteger) do |channels|
      stress_args.channels_per_server = channels
    end
    opts.on('--num_stubs_per_channel [INT]', OptionParser::DecimalInteger) do |stubs|
      stress_args.concurrent_calls = stubs
    end
    opts.on('--metrics_port [port]') do |port|
      stress_args.metrics_port = port
    end
  end.parse!
  stress_args
end

def main
  opts = parse_stress_args
  start(opts)
end

if __FILE__ == $0
  main
end
